core/src/toga/window.py

```diff
@@ -263,6 +263,13 @@ def close(self):
         self.app.windows -= self
         self._impl.close()
 
+    async def redraw(self):
+        """Request a redraw of the app, waiting until that redraw has completed."""
```

Comment:
I'm not sure about adding this to the public API, as it's unlikely to be useful outside of the Toga testbed. And as we mentioned the other day, whenever a property changes we should refresh things as necessary, just like the HTML DOM does. So any time an app needs to call `refresh` manually, that means there's a bug in Toga.

Instead, how about adding this as a method on the probe class? Then it could be implemented at either the window level or the widget level, depending how redrawing works on each platform.

Reply:
I'm not sure that we *would* want to force a refresh after every update - at least, not without also adding another mechanism to suspend those updates. 

Consider the case where you're updating a whole lot of widgets in a single handler - say - a big HUD style GUI with 100 widgets. At present, you can update each of the 100 widgets, and they'll all get refreshed when the handler exits and the redraw loop happens. If we bake a redraw into every widget change, there's going to be 99 "in-progress" redraws before the final update and redraw. Aside from the potential for drawing the UI in an "incomplete" state, that's going to be a pretty heavy redraw load. GUI paint operations aren't cheap, and they're the most likely to be observed by users as a "performance problem", because they'll manifest as graphical glitches you can clearly see.

The only way this approach would work in practice is if we also added a "pause updates" context manager that instructs Toga to *not* do per-widget updates while in the graphical context - this sort of thing is reasonably common for graphical APIs, so it wouldn't be completely weird; but it means decorating your code with "pause" wrappers anywhere that you've got moderately complex GUI updates.

The alternative is what I've got here. By default, redraws happen "as soon as practical". While I agree that this is somewhat contrary to a beginners expectation, it's also something that is reasonably easy to explain; and I'd argue that in most cases, it's the behavior you actually want, because it is optimised for visual performance. However, if you  need the GUI to update *right now*, `await redraw()` is an explicit way of doing the `await asyncio.sleep(0.01)` workaround. It's not something you need in the common case, but it's there if you do need it, and the default behaviour is semi-optimized.

Reply:
Following discussion in person; an async `redraw()` coroutine won't actually be that useful in practice, as the most likely use case is someone wanting to update in the middle of a synchronous handler; and the real fix for that is to not use a synchronous handler. On that basis, I'll demote redraw() to being a probe method.